### ISSUE TITLE
Location of application executable inside AppDir

### ISSUE THREAD

**Issue by ubruhin** [probonopd/linuxdeployqt#65]
Hi, 

my Qt application [LibrePCB](https://github.com/LibrePCB/LibrePCB) requires to load some resource files at runtime (for example some license templates) which I consider to store in the directory `../share/librepcb/` relative to the executable. So If the executable is installed to `/usr/bin/librepcb`, it searches for resources at `/usr/share/librepcb/`.

To make this working inside AppImages, the executable needs to be located in a subdirectory (e.g. `./bin/`) of the AppDir. Is there a way to do this with `linuxdeployqt`? As far as I can see, `linuxdeployqt` assumes that the application executable is located in the root of the AppDir...

**Comment by probonopd** [comment]
Hi @ubruhin. Thanks for your interest in linuydeployqt. Loading resources from locations relative to the main binary like you describe is a very good approach.

In fact, I started to work on a [FHS-like mode](https://github.com/probonopd/linuxdeployqt/issues/14) where linuxdeployqt should check whether the main executable is in a FHS-like structure, i.e., whether its parent directory is called `bin/` and (possibly) whether there is a `../lib/` directory. If so, it should use those directories to deploy to rather than putting `lib/` next to itself (which would result in a rather nonstandard `bin/lib/`).

My initial code on this can be found in the `feature/fhs` branch but it seems i am running into [this](https://github.com/probonopd/linuxdeployqt/issues/47) issue which I have been unable to resolve so far. Maybe you can find the culprit?

**Comment by ubruhin** [comment]
> In fact, I started to work on a FHS-like mode

Great to hear!

> Maybe you can find the culprit?

I have just looked into your source code, but unfortunately it's kind of hard to understand because it looks not very well structured (only 2 source files, no object oriented design, ...) :wink: Maybe I could do some refactoring first, if you agree with that?

**Comment by probonopd** [comment]
To be honest, I find the current structure hard to work with too, but I had modeled `linuxdeployqt` 1:1 after the already-existing `macdeployqt` which is part of Qt, in order to 
* benefit from the work that was already there in `macdeployqt`
* increase chances to have it included in Qt one day, too

I'm not against refactoring, just pointing out why it currently looks as it is now.

**Comment by ubruhin** [comment]
> I had modeled linuxdeployqt 1:1 after the already-existing macdeployqt

Alright, that's probably a good reason to keep the current structure.

> increase chances to have it included in Qt one day, too

That would be amazing! :smiley:  But wouldn't it be easier to bring `linuxdeployqt` into Qt if it doesn't know `AppImageKit` ( -> get rid of this dependency, because `AppImageKit` has nothing to do with Qt)? IMHO it would be no problem if developers need to execute two steps for creating AppImages (first `linuxdeployqt`, then `AppImageKit`).

> Also look at macdeployqt and windeployqt.

I will take a look on them the next few days.

**Comment by probonopd** [comment]
AppImageKit is not a dependency. It's a "works-well-with" relationship. It's not being used by default. I like to keep it.

**Comment by probonopd** [comment]
This is way over my head (this being my first Qt project ever), but I'd welcome this happening.

**Comment by probonopd** [comment]
FHS-like mode is now implemented, please try https://github.com/probonopd/linuxdeployqt/releases/tag/continuous

**Comment by ubruhin** [comment]
Now I get other errors:
- With Qt5.2 from Ubuntu Repos: https://travis-ci.org/LibrePCB/LibrePCB/jobs/201952944#L2495
- With Qt5.7 from beineri PPA: https://travis-ci.org/LibrePCB/LibrePCB/jobs/201952945#L2191

**Comment by probonopd** [comment]
Sorry for the trouble you are having @ubruhin 

> With Qt5.2 from Ubuntu Repos: https://travis-ci.org/LibrePCB/LibrePCB/jobs/201952944#L2495
```
ERROR: Could not start patchelf.
ERROR: Make sure it is installed on your $PATH, e.g., in /usr/local/bin.
ERROR: You can get it from https://nixos.org/patchelf.html.
```

> With Qt5.7 from beineri PPA: https://travis-ci.org/LibrePCB/LibrePCB/jobs/201952945#L2191

```
linuxdeployqt: /opt/qt57/lib/libQt5Core.so.5: version `Qt_5.8' not found (required by linuxdeployqt)
```

I need to investigate what is going on. Any help appreciated.

**Comment by ubruhin** [comment]
> Sorry for the trouble you are having @ubruhin

No problem ;)

> Are you sure you have the correct Qt on your $PATH as pointed out in README.md?

Yes, this is be done with sourcing the file `/opt/qt57/bin/qt57-env.sh`. If that wouldn't work, I couldn't build my own application too.

I will try to debug this on my local computer...

**Comment by probonopd** [comment]
@ubruhin please retest, I think the 

```
ERROR: Could not start patchelf.
ERROR: Make sure it is installed on your $PATH, e.g., in /usr/local/bin.
ERROR: You can get it from https://nixos.org/patchelf.html.
```
should be solved by your PR, https://github.com/probonopd/linuxdeployqt/commit/b3b9ec7d85ecc01a470e754fdca5754bca11a8bf

**Comment by ubruhin** [comment]
Yes, `Could not start patchelf` is fixed. But the `version Qt_5.8 not found` is still there.

**Comment by probonopd** [comment]
> The version Qt_5.8 not found is still there when using Qt5.7...

Is this when you run linuxdeployqt from its AppImage? I have compiled linuxdeployqt with Qt 5.8 and it needs that to run. The linuxdepoyqt.AppImage should contain that bundled...

**Comment by ubruhin** [comment]
> Where are you getting your linuxdeployqt from?

I use the one contained in the `linuxdepoyqt.AppImage`: https://github.com/LibrePCB/LibrePCB/blob/deploy_appimage_sandbox/dev/travis/install.sh#L18

**Comment by probonopd** [comment]
That is very strange. Can you please check with

```
wget -c "https://github.com/probonopd/linuxdeployqt/releases/download/continuous/linuxdeployqt-continuous-x86_64.AppImage"
chmod a+x linuxdeployqt-continuous-x86_64.AppImage 
./linuxdeployqt-continuous-x86_64.AppImage --appimage-extract
ldd ./squashfs-root/usr/bin/linuxdeployqt | grep libQt
```

It should load `libQt5Core.so.5` from within the AppImage, and for me it seems to do that:

```
me@host:~$ ldd ./squashfs-root/usr/bin/linuxdeployqt | grep libQt
	libQt5Core.so.5 => /home/me/./squashfs-root/usr/bin/../lib/libQt5Core.so.5 (0x00007ff08df2e000)
```

**Comment by ubruhin** [comment]
With Qt5.2 from the Ubuntu repository it [looks good](https://travis-ci.org/LibrePCB/LibrePCB/jobs/202990030#L579) to me:
```
libQt5Core.so.5 => /home/travis/build/LibrePCB/LibrePCB/./squashfs-root/usr/bin/../lib/libQt5Core.so.5 (0x00007f2a66679000)
```

But with Qt5.7 from the beineri PPA, it [looks bad](https://travis-ci.org/LibrePCB/LibrePCB/jobs/202990031#L479):
```
libQt5Core.so.5 => /opt/qt57/lib/libQt5Core.so.5 (0x00007fb265a41000)
```

I think this is somehow caused by [sourcing the file `/opt/qt57/bin/qt57-env.sh`](https://github.com/LibrePCB/LibrePCB/blob/deploy_appimage_sandbox/dev/travis/install.sh#L11), but what happens exactly?

**Comment by probonopd** [comment]
Please try to unset the environment variables set by `/opt/qt57/bin/qt57-env.sh` prior to running `linuxdeployqt`, with the only exception of `$PATH` that still should contain `qmake` from `/opt/qt57`.

**Comment by ubruhin** [comment]
That works!

I just saw that `/opt/qt57/bin/qt57-env.sh` contains following lines:

```
if [[ $(uname -m) == "x86_64" ]]; then
  export LD_LIBRARY_PATH=$QT_BASE_DIR/lib/x86_64-linux-gnu:$QT_BASE_DIR/lib:$LD_LIBRARY_PATH
else
  export LD_LIBRARY_PATH=$QT_BASE_DIR/lib/i386-linux-gnu:$QT_BASE_DIR/lib:$LD_LIBRARY_PATH
fi
```

I guess this is the problem. As the directories contained in `LD_LIBRARY_PATH` have highest priority when searching for libraries, libraries contained in the AppImage may be overridden.

But how to deal with that problem? ;)

**Comment by probonopd** [comment]
I think the clue may be in http://blog.qt.io/blog/2011/10/28/rpath-and-runpath/

> Some experiments and reading the glibc source code (check _dl_map_object in elf/dl-load.c) indicate that this is the library search order:
> 
```
Unless loading object has RUNPATH:
    RPATH of the loading object,
        then the RPATH of its loader (unless it has a RUNPATH), ...,
        until the end of the chain, which is either the executable
        or an object loaded by dlopen
    Unless executable has RUNPATH:
        RPATH of the executable
LD_LIBRARY_PATH
RUNPATH of the loading object
ld.so.cache
default dirs
```
